keps/sig-storage/361-local-ephemeral-storage-isolation/README.md

```diff
@@ -0,0 +1,490 @@
+  emptyDir volume) plus containers disk usage should not exceed the pod-level resource requirement
+  (the sum of all container’s limit). 
+* EmptyDir SizeLimit: Because emptyDir volume is a pod-level resource which is managed separately,
+  we also add a sizeLimit for emptyDir Volume for additional storage isolation. If this limit is
+  set for emptyDir volume (default medium), eviction manager will validate this limit with emptyDir
+  usage too in additional to the above pod-level resource constraints. 
+
+### Eviction Policy and Scheduler Predicates
+
+CPU and memory use cgroup for limiting the resource usage. However, cgroup for disk usage is not
+available. Our current design is to evict pods when exceeding the limit set for local storage.
+The eviction policy is listed as follows.
+
+* If the container writable layer (overlay) usage exceeds its limit set for this container, pod gets evicted.
+* If the emptyDir volume usage exceeds the sizeLimit, pod gets evicted.
+* If the sum of the usage from emptyDir and all contains exceeds the sum of the container’s local storage
+  limits (pod-level limit), pod gets evicted.
+
+Pod Admission
+
+* When scheduler admits pods, it sums up the storage requests from each container and pod can be scheduled
+  only if the sum is smaller than the allocatable of the local storage space.
+
+Resource Quota
+
+This feature adds two more resource quotas for storage. The request and limit set constraints on the
+total requests/limits of all containers’ in a namespace.
+
+* requests.ephemeral-storage
+* limits.ephemeral-storage
+
+LimitRange 
+
+* Similar to CPU and memory, admin could use LimitRange to set default container’s local storage
+  request/limit, and/or minimum/maximum resource constraints for a namespace. 
+
+Node Allocatable Resources
+
+* Similar to CPU and memory, ephemeral-storage may be specified to reserve for kubelet or system.
+  example,
+  --system-reserved=[cpu=100m][,][memory=100Mi][,][ephemeral-storage=1Gi][,][pid=1000]
+  --kube-reserved=[cpu=100m][,][memory=100Mi][,][ephemeral-storage=1Gi][,][pid=1000]
+
+### Risks and Mitigations
+
+* This feature introduces CPU and IO overhead due to the monitoring of disk usage.
+
+* Pod will be evited when resource limit is exceeded. This might not be the ideal
+  way of handling resource over use. 
+
+* Before Pod is evicted, resource usage might temporally exceed the limit.
+
+* Not all local storage is tracked by this feature. For example, [CSI ephmeral
+  volumes](https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#csi-ephemeral-volumes),
+  if they use node resources, will (probably) be accumulated into system
+  resources. A KEP has been proposed in this area
+  ([#1472](https://github.com/kubernetes/enhancements/issues/1472)). Details
+  like CSI ephemeral volumes which share local storage resources across pods
+  have not been addressed. We feel this is best tracked with the CSI ephemeral
+  volumes feature.
+
+## Design Details
```

Comment:
I'm surprised not to see more details here - how will we get the current usage?  What exactly is included and not included in that total?  if I have multiple `emptyDir` volumes are they guaranteed to share a disk?  How often will this be polled and why do we think that is good enough?  How much extra IO and CPU will that use?  Are we depending on kernel's quota subsystem?  Why or why not?

Reply:
+1 for all of that

And wearing my scalability/PRR hat, I would like to understand better the IO and CPU overhead mentioned here - how big that is?